Sources/Workspace/Workspace.swift

```diff
@@ -836,7 +844,10 @@ extension Workspace {
         diagnostics: DiagnosticsEngine
     ) -> [Manifest] {
         let rootManifests = packages.compactMap({ package -> Manifest? in
-            loadManifest(packagePath: package, url: package.pathString, packageKind: .root, diagnostics: diagnostics)
+            delegate?.willLoadPackage(url: package.pathString, kind: .root)
```

Comment:
nit: `willLoadManifest` is probably slightly more appropriate since package loading currently refers to the operation that scans the file system for building up the package model.

Reply:
You're correct, of course, in terms of what is actually going on between the _will_ and the _did_.  And I started that way here, and called the callback whenever we loaded a package manifest, but it's not actually what a client wants (at least not from a high-level perspective).

The broader perspective is this:  the current delegate callbacks are very "technical" in that they reflect the steps that libSwiftPM is taking.  But what a client often wants is the conceptual view, from which it can form things like a user-centric log, etc.  In the case of packages, libSwiftPM may actually call the manifest loading code multiple times for a package, while to a user, the package is only used once.

In the past, clients (or the one I've worked on) have kept side data inferred from the existing set of callbacks, but what I'd like to do is to evolve the set of callbacks into something that is more suited to what clients need, and then find where in libSwiftPM to put in the hooks rather than the other way around.

There could of course be better ways to do this, but that's the motivation here.

Reply:
I agree with the goal here, I was just commenting on the name of the delegate since swiftpm uses "package loading" to refer to the operations in https://github.com/apple/swift-package-manager/tree/master/Sources/PackageLoading, which is not what's happening here so it can become confusing to other maintainers. Maybe we can find a different name to reflect the intent for libSwiftPM clients here.

Reply:
Ah got it.  Right, that makes sense.  I'll rename it, or if there's a better place to hook in to convey the conceptual event of "loading a package", that would be even better.  I'll take a look at the source file you mentioned (still getting familiar with several parts of the code base).

Reply:
Also, I should say (for anyone reading) that I realize that the full list of loaded packages can be retrieved by the client after the entire package graph operation finishes, but the idea here would be to get those events in real-time (since package repository fetching can take a lot of time).

Reply:
Actually, a related question:  if there's a manifest loading failure, then the callbacks for any "downstream" packages don't get called.  So maybe I'm thinking about this the wrong way.  Basically from a client side, I'd like to get told whenever a new package is first "seen" in a workspace, and to get any diagnostics for its manifest as soon as they are known (even as the result of the operation, with any possible lengthy repo fetching, goes on).  Getting told about loading of the manifest as it's happening seems to be pretty good proxy, but maybe there are better ones.